### PR TITLE
[AMDGPU] Preserve `noundef` and `range` during kernel argument loads

### PR DESCRIPTION
**Pull request by krzysz00**
This commit ensures than noundef (which is frequently a prerequisite for other annotations) and range() annotations on kernel arguments are copied onto their corresponding load from the kernel argument structure.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/lower-kernargs.ll [hunk 1]
```diff
@@ -255,6 +255,25 @@ define amdgpu_kernel void @kern_i32(i32 %arg0) {
...
+; MESA-NEXT:    [[KERN_RANGE_NOUNDEF_I32_KERNARG_SEGMENT:%.*]] = call nonnull align 16 dereferenceable(260) ptr addrspace(4) @llvm.amdgcn.kernarg.segment.ptr()
+; MESA-NEXT:    [[ARG0_KERNARG_OFFSET:%.*]] = getelementptr inbounds i8, ptr addrspace(4) [[KERN_RANGE_NOUNDEF_I32_KERNARG_SEGMENT]], i64 36
+; MESA-NEXT:    [[ARG0_LOAD:%.*]] = load i32, ptr addrspace(4) [[ARG0_KERNARG_OFFSET]], align 4, !range [[RNG2:![0-9]+]], !invariant.load [[META1]], !noundef [[META1]]
+; MESA-NEXT:    store volatile i32 [[ARG0_LOAD]], ptr addrspace(1) poison, align 4
+; MESA-NEXT:    ret void
+;
+  store volatile i32 %arg0, ptr addrspace(1) poison
+  ret void
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/lower-kernargs.ll:276]: Add a noundef test that covers FP, vector, pointer

File: llvm/test/CodeGen/AMDGPU/lower-kernargs.ll [hunk 2]
```diff
@@ -255,6 +255,25 @@ define amdgpu_kernel void @kern_i32(i32 %arg0) {
...
+; HSA-NEXT:    ret void
+;
+; MESA-LABEL: @kern_range_noundef_i32(
+; MESA-NEXT:    [[KERN_RANGE_NOUNDEF_I32_KERNARG_SEGMENT:%.*]] = call nonnull align 16 dereferenceable(260) ptr addrspace(4) @llvm.amdgcn.kernarg.segment.ptr()
+; MESA-NEXT:    [[ARG0_KERNARG_OFFSET:%.*]] = getelementptr inbounds i8, ptr addrspace(4) [[KERN_RANGE_NOUNDEF_I32_KERNARG_SEGMENT]], i64 36
+; MESA-NEXT:    [[ARG0_LOAD:%.*]] = load i32, ptr addrspace(4) [[ARG0_KERNARG_OFFSET]], align 4, !range [[RNG2:![0-9]+]], !invariant.load [[META1]], !noundef [[META1]]
+; MESA-NEXT:    store volatile i32 [[ARG0_LOAD]], ptr addrspace(1) poison, align 4
+; MESA-NEXT:    ret void
+;
+  store volatile i32 %arg0, ptr addrspace(1) poison
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/lower-kernargs.ll:273]: Avoid store to poison

Comment (krzysz00) [reply llvm/test/CodeGen/AMDGPU/lower-kernargs.ll:273]: Is there a good idiom here for a side effect that keeps the argument alive? I remember someone recently complaining about stores to undef

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/lower-kernargs.ll:273]: store to null in non-0 address space is technically well defined, but probably best to just use an ordinary function argument or global variable declaration

**Review by arsenm — APPROVED** [review]

**Pull request merged**